### PR TITLE
refactor: extract position-loss decay weights helper

### LINKED ISSUE #12: refactor: extract a dedicated ploss weighting function
## Problem

The TTT position loss weighting logic `0.8**i` is hardcoded and duplicated in 3 separate locations in `eagle3_trainer.py`:

- Line 251 (`_backward`): `ploss_weight = [0.8**i for i in range(len(plosses))]`
- Line 302 (`_aggregate_eval_metrics`): `[0.8**i for i in range(avg_plosses.shape[0])]`
- Line 394 (`_aggregate_metrics`): `[0.8**i for i in range(avg_plosses.shape[0])]`

The weighting strategy, decay factor, and reduction are all implicitly embedded in each call site with no abstraction.

## Proposed Solution

Extract a `ploss` function that takes only the per-position losses and fully encapsulates the weighting strategy internally:

```python
def weighted_ploss(plosses: list[torch.Tensor]) -> torch.Tensor:
    """Compute weighted position loss.

    Encapsulates the full weighting strategy — callers don't need to know
    whether it's exponential decay, linear decay, or something else.
    """
    weights = [0.8 ** i for i in range(len(plosses))]
    return sum(w * p for w, p in zip(weights, plosses)) / sum(weights)
```

The key design choice: the function signature is just `plosses` in, scalar out. The weighting strategy is an implementation detail hidden inside. This makes it easy to swap to a completely different strategy (e.g., non-uniform, learned weights, position-dependent) without changing any call sites.

This function should be used in:
- `_backward` for the training loss
- `_aggregate_metrics` for the training metric
- `_aggregate_eval_metrics` for the eval metric

## Files

- `torchspec/training/eagle3_trainer.py`

### PR DESCRIPTION
**Pull request by ppraneth**
Closes #12.

  ## What this does
  - Pulls `[0.8**i for i in range(n)]` into a helper `_position_decay_weights(n)` in
  `eagle3_trainer.py`.
  - Calls it from the three spots that had the list comprehension inlined: `_backward`,
  `_aggregate_metrics`, `_aggregate_eval_metrics`.
  - Drops a now-redundant comment (`# Compute weighted loss matching _backward's 0.8^i weighting`) —
  the shared helper makes that link explicit.

  ## Comparision to original issue:
  The issue proposed one helper of the form `sum(w*p) / sum(w)` for all three sites. I didn't do that
  Reason:

  - The three sites don't compute the same reduction today.
    - `_backward` → weighted **sum**: `Σ wᵢ·pᵢ / accumulation_steps` (no `/ sum(w)`).
    - `_aggregate_metrics` and `_aggregate_eval_metrics` → weighted **average**: `Σ wᵢ·vᵢ / sum(w)`.
  -  They differ by a factor of `Σ 0.8^i`, which depends on `ttt_length`:
    - `ttt_length=4` → ~2.95
    - `ttt_length=7` (default in most configs) → ~3.95  
    - So I only shared the **weights**. Each reduction stays visible at its call site.

  ## Open question (not for this PR)
  - Should `_backward` also normalize by `sum(w)`?

### CODE REVIEW

**Comment by ppraneth** [comment]
@yubofredwang can you please review this pr?

**Comment by yubofredwang** [comment]
@ppraneth overall looks good to me, along this fix, can we actually add an option to adjust the ratio of the decay instead of fixing at 0.8? 

for example, user can input an array [0.8, 0.9, 0.9]

**Comment by ppraneth** [comment]
@yubofredwang 
Happy to add this. Just to make sure I match what you want — do you mean:

(a) the user supplies the full weight array directly (e.g. `ploss_weights: [1.0, 0.8, 0.64, ...]`), or
(b) the user specifies the decay base (e.g. `ploss_decay: 0.9` → `[0.9**i for i in range(n)]`)?

Your example reads more like (a), but I wanted to confirm before I implement it.

**Comment by yubofredwang** [comment]
I think (a), just let user figure out the actual weight by themselves. Thanks!

**Comment by ppraneth** [comment]
@yubofredwang Made the changes can you review it again?
